dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/DefaultTrackedEntityInstanceService.java

```diff
@@ -248,19 +248,19 @@ public List<Long> getTrackedEntityInstanceIds( TrackedEntityInstanceQueryParams
      */
     private void handleSortAttributes( TrackedEntityInstanceQueryParams params )
     {
-        if ( params.hasAttributeAsOrder() )
-        {
-            // Collecting TEAs for all non static sort order columns.
-            List<TrackedEntityAttribute> sortAttributes = params.getOrders().stream()
-                .filter( orderParam -> !OrderColumn.isStaticColumn( orderParam.getField() ) ).map( orderParam -> {
-                    return attributeService.getTrackedEntityAttribute( orderParam.getField() );
-                } ).collect( Collectors.toList() );
+        List<String> ordersIdentifier = params.getOrders().stream()
+            .map( OrderParam::getField )
+            .collect( Collectors.toList() );
+        List<QueryItem> sortableAttributes = params.getAttributesAndFilters().stream()
+            .filter( queryItem -> ordersIdentifier.contains( queryItem.getItemId() ) )
+            .collect( Collectors.toList() );
 
-            // adding to attributes conditionally if they are also not present
-            // in filters.
-            params.addAttributesIfNotExist( QueryItem.getQueryItems( sortAttributes ).stream()
-                .filter( sAtt -> !params.getFilters().contains( sAtt ) ).collect( Collectors.toList() ) );
-        }
+        List<TrackedEntityAttribute> sortAttributes = sortableAttributes.stream()
```

Comment:
we only use ordersIdentifier and sortableAttributes once, could we avoid declaring these two lists and then map, filter and map all in one statement?